### ISSUE TITLE
Redirect to any path of current domain

### ISSUE THREAD

**Issue by samuelsolis** [agentrickard/domain#203]
I don't know way if I set a Form redirect of this way:

> $form_state->setRedirect('mymodule.mypathname');

The redirection aren't correct, it generate me an external url and Drupal throw me this error:

> Redirects to external URLs are not allowed by default, use \Drupal\Core\Routing\TrustedRedirectResponse for it.

I was debugging and the url generator return the canonical domain, but I wan't to redirect to current domain with this path.

I tested using domains and domain alias and both result are the same.

It is a bug? Am I doing somthing wrong?

Thank in advance.

**Comment by samuelsolis** [comment]
I have to add that my $settings['trusted_host_patterns']  are correctly configured, it contents all domain that are defined in the domain list.

**Comment by agentrickard** [comment]
Can you share a link to the code, or at least the route definition? 

I don't see this happening with core forms (e.g. `admin/config/regional/date-time/formats/manage/long`) or with the DomainForm itself, where we use `$form_state->setRedirect('domain.admin');`.

Are you using Domain Source and routing back to a node page? In that case, it may rewrite the URL as external.

**Comment by samuelsolis** [comment]
Currently this code are running ok, but it isn't very elegant:
http://pastebin.com/rZFHkwjR

Domain Source is not installed, it isn't necessary for me rigth now I think. As you say, it's probably a bug my domain configuration.

While I was answering you, in order to recollect more infomation and following your suggestion, I made another tests and I discover that the error is in all Drupal redirection if I not in the main domain: 

- Saving domain config Form using an alternative domain (not the main domain) and the error is the same exception.
- Making a loggin (in a alternative domain).
- etc.

How can I debug this? What domain config are wrong?

This is my domain.settings.yml:
http://pastebin.com/WX3MgWz8

And 2 domain settings in yalm too:
The main domain:
http://pastebin.com/hCUEwDSB
a secondary domain:
http://pastebin.com/awx76uy1

How can I debug this? Can you give me any track or guideline for resolving it?

Thank in advance. :)

**Comment by agentrickard** [comment]
Well, the domain.settings.yml config isn't actually used yet, so you can ignore that. These YML files look fine.

Do you have Aliases for the domains? Are they set to redirect?

Are the domains set to "active" status? If not, does your user account have the permission to view inactive domains? 

There are two obvious things that could be happening:

# Domain Alias is issuing a redirect. You could check the logic in `domain_alias_domain_request_alter()`
# The use of an inactive domain is causing a redirect. See `Drupal\domain\EventSubscriber\DomainSubscriber`

The only other thing I could see is if it's still possible to set a base_url in services.yml. Don't do that.

**Comment by agentrickard** [comment]
Sorry, I didn't look at the code in the first link. This has nothing to do with Domain, and is about how you are telling Drupal to send the redirect. setRedirect() only accepts paths. I think you need setResponse() -- see below.

A few things:

* It looks like the is using older code. The 8.x-1.x has had some changes, so update that.
* What is $this->domain in this context? It needs to be the DomainNegotirator class getActiveDomain() method..
* Use the new $domain->buildUrll($path) method and make sure you have an active domain.

```
    //Redirect to thanks page.
    //@todo refactor this when we can do an internal redirect with domain.
    $current = \Drupal::service('domain.negotiator')->getActiveDomain();
    if(isset($response->result) && $current) {
      $url = Url::fromUri($current->buildUrl('/landing-register/thanks'));
    } else {
      drupal_set_message($this->t('Something is wrong. Please, contact with an administrator.'), 'error');
      $url = Url::fromUri($current->buildUrl('/landing-register/thanks'));
    }
    $form_state->setRedirectUrl($url);
```

* When setting a redirect to an absolute URL, we must use a TrustedRedirectResponse. I don't think that's available to the FormState, so you might have to use $form_state->setResponse() instead.

```
  use Drupal\Core\Routing\TrustedRedirectResponse;
   $response = new TrustedRedirectResponse($current->buildUrl('/landing-register/thanks'), 301);
  $form_state->setResponse($response);   

```

The TrustedRedirectReponse should work in that case. If I understand the code correctly, however, it would potentially break this form if used in an AJAX context.

The other option is to set up a new Route that takes a {domain} and {path} argument and redirect to that and issue the TrustedRedirectResponse from your own controller.

**Comment by agentrickard** [comment]
Closing, though I'd love to know if you solved this issue.

**Comment by samuelsolis** [comment]
I didn't have enough time to debug it, but I will do it asap. I write here if I resolve it, the problem, and the solutions. 

Thanks for your help, :-)

**Comment by tekNorah** [comment]
My client is using the Open Social distribution in combination with the domain module to configure several subdomains that point to specific Community Groups, acting as "parent" groups.  I have created subgroup functionality using the domain_entity module to create a relationship between the subgroup and the subdomain pointing to a specific Community Group.

I have setup business rules to automatically assign a new Open/Closed Group to the current domain.

I have also setup aliases for each environment.  A local dev using docksal & a staging environment on the same servers as production (Cloudways).
 
The last feature request I am working on deals with redirecting a group to the appropriate domain.  So, in a custom module, I created \Drupal\my_module\EventSubscriber\RedirectSubscriber and in the public function checkForRedirection, I have setup a redirect that does logic to discover if the group/subgroup is on the wrong domain and build a url, dynamically based on this logic.

For example, If a user goes to a subgroup via https://green.project.docksal/groups/100/stream and it turns out that that the subgroup is assigned to the the blue subdomain, my logic will build the url: https://blue.project.docksal/groups/100/stream and redirect using:
`$event->setResponse(new TrustedRedirectResponse($group_domain_url));`

Now, this **works** on my local dev env, but when I push up to the staging Env on Cloudways, it produces the error:
`ERR_TOO_MANY_REDIRECTS`

The trusted host pattern in the settings file is determined by:
```
$httphost = $_SERVER['HTTP_HOST'];
$settings['trusted_host_patterns'] = [
  '^' . $httphost . '$',
];
```

Any ideas on how I can troubleshoot/solve this problem?
